src/badguy/corrupted_granito_big.cpp

```diff
@@ -0,0 +1,170 @@
+
+  bool crack = try_cracking();
+  if (!crack && !m_crack_timer.paused() && m_crack_timer.started()) m_crack_timer.pause();
+
+  if (m_shake_timer.started())
+    m_shake_delta = static_cast<float>(graphicsRandom.rand(-3, 3));
+  else
+    m_shake_delta = 0.f;
+}
+
+void CorruptedGranitoBig::draw(DrawingContext &context)
+{
+  m_sprite->draw(context.color(), {get_pos().x + m_shake_delta, get_pos().y}, m_layer, m_flip);
+}
+
+void CorruptedGranitoBig::kill_fall()
+{
+  m_state = STATE_BROKEN;
+  set_action("broken", m_dir);
+  set_colgroup_active(COLGROUP_MOVING_ONLY_STATIC);
+  m_col.set_unisolid(false);
+
+  run_dead_script();
+
+  Sector::get().add<Shard>(get_bbox().get_middle(), Vector(100.f, -500.f), SHARD_SPRITE);
+  Sector::get().add<Shard>(get_bbox().get_middle(), Vector(270.f, -350.f), SHARD_SPRITE);
+  Sector::get().add<Shard>(get_bbox().get_middle(), Vector(-100.f, -500.f),SHARD_SPRITE);
+  Sector::get().add<Shard>(get_bbox().get_middle(), Vector(-270.f, -350.f),SHARD_SPRITE);
+
+  crack_effects(6);
+}
+
+bool CorruptedGranitoBig::try_cracking()
+{
+  using RaycastResult = CollisionSystem::RaycastResult;
+
+  Player* plr = get_nearest_player();
+  if (!plr) return false;
+
+  Vector mid = get_bbox().get_middle();
+  Vector plrmid = plr->get_bbox().get_middle();
+
+  RaycastResult result = Sector::get().get_first_line_intersection(get_bbox().get_middle(),
+                                                                   plr->get_bbox().get_middle(),
+                                                                   false,
+                                                                   get_collision_object());
+
+  auto* obj = std::get_if<CollisionObject*>(&result.hit);
+  if (!obj || *obj != plr->get_collision_object()) return false;
+
+  if (glm::distance(mid, plrmid) > 32.f*RANGE) return false;
+
+  crack();
+  return true;
+}
+
+void CorruptedGranitoBig::crack()
+{
+  if (m_state == STATE_CRACK2 && m_crack_timer.check())
+  {
+    kill_fall();
+    return;
+  }
+
+  if (m_crack_timer.paused())
+  {
+    m_crack_timer.unpause();
+  }
+  else if (!m_crack_timer.started())
+    m_crack_timer.start(CRACK_TIME);
+
+  float progress = m_crack_timer.get_timegone() / m_crack_timer.get_period();
+
+  if (m_state == STATE_CRACK1 && progress >= 0.5f)
+  {
+    m_state = STATE_CRACK2;
+    set_action("cracked-2", m_dir);
+    crack_effects(4);
+  }
+  else if (m_state == STATE_READY)
+  {
+    m_state = STATE_CRACK1;
+    set_action("cracked-1", m_dir);
+    crack_effects(3);
+  }
+}
+
+void CorruptedGranitoBig::crack_effects(size_t particles)
+{
+  SoundManager::current()->play("sounds/brick.wav", get_pos());
+  m_shake_timer.start(SHAKE_TIME);
+
+  const auto gravity = Sector::get().get_gravity() * 100;
+  std::vector<std::string> pieces = {"piece-0", "piece-1", "piece-2", "piece-3", "piece-4", "piece-5"};
+  pieces = std::vector<std::string>(pieces.begin(), pieces.begin() + particles);
+  for (std::string action : pieces)
```

Comment:
This can be replaced with `const std::vector` and normal for: `for (size_t i = 0; i<particles; i++)`.